hazelcast/src/main/java/com/hazelcast/cache/impl/operation/LocalMergeOperation.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl.operation;
+
+import com.hazelcast.cache.CacheEntryView;
+import com.hazelcast.cache.impl.CacheDataSerializerHook;
+import com.hazelcast.cache.impl.CacheRecordStore;
+import com.hazelcast.cache.impl.ICacheRecordStore;
+import com.hazelcast.cache.impl.ICacheService;
+import com.hazelcast.cache.impl.event.CacheWanEventPublisher;
+import com.hazelcast.cache.impl.record.CacheRecord;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.BackupAwareOperation;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.ServiceNamespace;
+import com.hazelcast.spi.ServiceNamespaceAware;
+import com.hazelcast.spi.SplitBrainMergeEntryView;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+import com.hazelcast.spi.impl.AbstractNamedOperation;
+import com.hazelcast.spi.serialization.SerializationService;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.cache.impl.CacheEntryViews.createDefaultEntryView;
+import static com.hazelcast.util.MapUtil.createHashMap;
+
+public class LocalMergeOperation extends AbstractNamedOperation implements BackupAwareOperation, ServiceNamespaceAware {
```

Comment:
The name `MergeOperation` is already taken for `ICache`, `IMap` and `ReplicatedMap`. So I chose `LocalMergeOperation` for all the new ones. They are normally executed just locally.

But you are right, in case of a migration (or any other case that the partition doesn't belong to that member), the operation is retried via `InvokeOnPartitions.retryFailedPartitions()` and `MergeOperationFactory.createPartitionOperation()`.

I thought of renaming the old operations to `LegacyMergeOperation` (including a rename of the IDS constants). This creates a bit more confusion in the PR, but maybe the cleaner solution.

Reply:
`LegacyMergeOperation` sounds good, in the next dev cycle it can be removed so it's not one of those `legacy` stuff that live forever :).
Also, please add a marker comment `// RU_COMPAT_3_9` so we know the class is still there just for 3.9 compatibility and can be removed in the next iteration.

Reply:
We can't remove it in 3.x, since the old merge policies are out in the wild and used in existing configurations. The only thing which can be removed is the cluster version check for the new merge policies. During an 3.10 to 3.11 RU the new merge policies are allowed to be used.

Reply:
Ah, right.